Train Response Selector on labels by default instead of text

dakshvar22 opened the pull request:
**Proposed changes**:
- Fixes https://github.com/RasaHQ/research/issues/111
- Adds `intent_response_key` attribute to `Message` class which will hold the complete `intent/response_key`(for example - `faq/languages`) label. This attribute is not dense featurizable.
- Added a flag - `use_text_as_label` in `ResponseSelector` to configure whether to train on text of responses or the `intent/response_key` label. Default being training on `intent/response_key` label.
- Sub-intents inside a retrieval intent(for example - `faq/languages` inside `faq`) can now have multiple response templates. These are handled by always using the first response template containing the `text` attribute for training and selecting a random template for bot utterance.
- Deprecate `full_retrieval_intent` from the output of the ResponseSelector and instead use `intent_response_key` to store the full intent name as it appears in the training data. This is more consistent with the naming schema used in the code.
- Deprecate `name` as the key to store the predicted response text and instead use `response_templates` to output all available response templates for the predicted `intent_response_key`.
- Added a key `all_retrieval_intents` to response selector output which will hold a list of all retrieval intents(`faq`,`chitchat`, etc.) that are present in the training data.
- So now the output looks like - 
```
"response_selector": {
    "all_retrieval_intents": ["faq"],
    "default": {
      "response": {
        "id": 1388783286124361986, "confidence": 1.0, "intent_response_key": "faq/is_legit",
        "response_templates": [
          {
            "text": "absolutely",
            "image": "https://i.imgur.com/nGF1K8f.jpg"
          },
          {
            "text": "I think so."
          }
        ],
      },
      "ranking": [
        {
          "id": 1388783286124361986,
          "confidence": 1.0,
          "intent_response_key": "faq/is_legit"
        },
      ]
```
- All response selector related evaluation artifacts - `report.json, successes.json, errors.json, confusion_matrix.png` now use the complete `intent_response_key` as the label instead of the actual response text. Makes them look more concise and clean and at the same time no information is lost.
- `response_selection_successes.json` looks like - 
```
{
      "text": "Is this really legit?",
      "intent_response_key_target": "faq/is_legit",
      "intent_response_key_prediction": {
              "name": "faq/is_legit",
              "confidence": 0.88
      }
}
```
- `response_selection_errors.json` looks like - 
```
{
      "text": "Is this really legit?",
      "intent_response_key_target": "faq/is_legit",
      "intent_response_key_prediction": {
              "name": "faq/ask_cost",
              "confidence": 0.6
      }
}
```


**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

dakshvar22 commented on rasa/nlu/selectors/response_selector.py:
```diff
@@ -204,6 +205,9 @@ def required_components(cls) -> List[Type[Component]]:
         RETRIEVAL_INTENT: None,
+        # Boolean flag to check if actual text of the response
+        # should be used as ground truth label for training the model.
+        TRAIN_ON_TEXT: True,
```
Reminder for myself: Change this back to False, before the merge. It is `True` right now for one run of model regression tests.

tabergma commented on rasa/nlu/test.py:
```diff
@@ -1333,15 +1332,17 @@ def get_eval_data(
                 response_prediction_key, {}
             ).get(OPEN_UTTERANCE_PREDICTION_KEY, {})
 
+            # print(response_prediction)
+
```
```suggestion
```

dakshvar22 replied:
Ahh, thanks for catching that.

tabergma commented on rasa/nlu/tokenizers/tokenizer.py:
```diff
@@ -100,15 +107,30 @@ def process(self, message: Message, **kwargs: Any) -> None:
 
-    def _split_intent(self, message: Message):
-        text = message.get(INTENT)
+    def _tokenize_on_split_symbol(self, text: TEXT) -> List[TEXT]:
```
```suggestion
    def _tokenize_on_split_symbol(self, text: Text) -> List[Text]:
```

dakshvar22 replied:
Raaiight, now I know why my type checks are failing 🙌

tabergma commented on rasa/nlu/training_data/training_data.py:
```diff
@@ -243,15 +239,22 @@ def _fill_response_phrases(self) -> None:
+
+                # Use the first response text as training label if needed downstream
+                for assistant_utterance in assistant_utterances:
+                    if assistant_utterance.get(TEXT, None):
```
```suggestion
                    if assistant_utterance.get(TEXT) is not None:
```
The default return value is already `None`.

tabergma commented on rasa/nlu/training_data/training_data.py:
```diff
@@ -421,7 +424,7 @@ def validate(self) -> None:
         # emit warnings for response intents without a response template
         for example in self.training_examples:
-            if example.get(RESPONSE_KEY_ATTRIBUTE):
+            if example.get(RESPONSE_KEY) and example.get(RESPONSE, None) is None:
```
```suggestion
            if example.get(RESPONSE_KEY) and example.get(RESPONSE) is None:
```

tabergma commented on rasa/nlu/selectors/response_selector.py:
```diff
@@ -313,20 +322,22 @@ def preprocess_train_data(self, training_data: TrainingData) -> RasaModelData:
-    def _full_response(self, label: Dict[Text, Any]) -> Optional[Dict[Text, Any]]:
+    def _full_response(
+        self, label: Dict[Text, Any]
+    ) -> Tuple[Any, List[Dict[Text, Any]]]:
```
not sure, but I think we can replace `Tuple[Any, List[Dict[Text, Any]]]` with `Tuple[Text, List[Dict[Text, Any]]]`

dakshvar22 replied:
I thought so too, but type check fails I think bc `label` as type `Dict[Text, Any]` which is valid because a value in that dict can be `None` as well.

tabergma replied:
Can we then change the type of `label` to `Dict[Text, Optional[Text]]`?

tabergma commented on rasa/nlu/selectors/response_selector.py:
```diff
@@ -366,9 +387,13 @@ def process(self, message: Message, **kwargs: Any) -> None:
 
         prediction_dict = {
-            "response": self._full_response(label) or {TEXT: label.get("name")},
+            "response": {
```
As we changed this, do we need to update the components page in our docs?

dakshvar22 replied:
Yes, I'll update the docs in a separate PR.

tabergma commented on rasa/nlu/selectors/response_selector.py:
```diff
@@ -366,9 +387,13 @@ def process(self, message: Message, **kwargs: Any) -> None:
         prediction_dict = {
-            "response": self._full_response(label) or {TEXT: label.get("name")},
+            "response": {
+                "id": label["id"],
```
just for my understanding, what exactly is the id of label?

dakshvar22 replied:
It's the hash of whatever is stored in the `name` of label.

tabergma commented on rasa/nlu/selectors/response_selector.py:
```diff
@@ -480,7 +505,11 @@ def _prepare_layers(self) -> None:
-        self._prepare_sequence_layers(self.label_name)
+
+        # When training on labels, don't create transformer layers for the label
+        self._prepare_sequence_layers(
```
just to clarify: the incoming user message is still going through the transformer. the responses are just going through the transformer if we are using the text, but not if we are using the label, in that case it is just a ffnn and an embedding layer, correct?

dakshvar22 replied:
Yes, that's correct. Just the transformer layer is skipped if training on labels.

tabergma commented on rasa/nlu/test.py:
```diff
@@ -451,15 +455,10 @@ def evaluate_response_selections(
         plot_utils.plot_confusion_matrix(
             confusion_matrix,
-            classes=_labels,
+            classes=labels,
```
So `labels` is now always the retrieval intent?

dakshvar22 replied:
Yes, it makes the files look much cleaner IMO and conveys all the information also anyways.

tabergma commented on rasa/nlu/selectors/response_selector.py:
```diff
@@ -298,12 +303,16 @@ def preprocess_train_data(self, training_data: TrainingData) -> RasaModelData:
                 "all retrieval intents."
             )
 
+        label_attribute = RESPONSE if self.train_on_text else INTENT_RESPONSE_KEY
```
Minor: Can we somehow unify the names `RESPONSE` and `INTENT_RESPONSE_KEY`? If you think it does not make sense it is also fine.

dakshvar22 replied:
But `label_attribute` is actually intended to unify them for the later part of the code, so they are just used here in this one line.

erohmensing commented:
> Move full_retrieval_intent key in the output of ResponseSelector inside the response key. 

I'm a bit concerned about this, as the `parse_data` that comes out of events could be considered a public API and moving this could break users' front ends (e.g. [rasa x](https://github.com/RasaHQ/rasa-x/pull/3488/files#diff-651c9f3a1a70660a008b93746c021e3fR442)), actions (e.g. [sara](https://github.com/RasaHQ/rasa-demo/blob/master/actions/actions.py#L228-L230) ), or analytics functions that consume events, etc. 

Ideally we'd have a way to deprecate this if we're moving it, but I'm not sure how one would go about that 🤔

tabergma reviewed (approved):
Looks good from my side 👍 . I don't have time to test it today, but I think Ella is doing it. Apart from that we should update the documentation and add also an example for the training data, but it is fine if that is addressed in a separate PR. 

Thanks for updating the response selector 💯

erohmensing commented on changelog/6466.improvement.rst:
```diff
@@ -0,0 +1,6 @@
+[Response selector](retrieval-actions.mdx) is now trained on retrieval intent labels by default instead of actual response text.
+
+If you want to use the actual response text for training, then use ``train_on_text=True`` in the configuration of ``ResponseSelector``.
```
Can we maintain the description of the features as described here in the changelog, but also mention in the migration guide _why_ this change was made (`For most models, this should improve training time of the component with similar confidence results.`, if that's accurate) as well as how to revert to the default? It'd be good to have this info whenever we change defaults that we know could actually be bad for peoples' models, e.g. tom's example of chitchat/faq_1, chitchat/faq_2. example below:

erohmensing replied:
[Response Selectors](retrieval-actions.mdx) are now trained on retrieval intent labels by default instead of actual response text. For most models, this should improve training time of the `ResponseSelector`, with similar confidence results.

If you want to revert to the pre-2.0 default behavior, add the `train_on_text: true` parameter to your `ResponseSelector` component.

dakshvar22 commented:
@erohmensing Hmm, good point. I mean it would be good to maintain compatibility but we don't have to maintain API compatibility for a new major version right, specially for an earlier experimental feature? We definitely need to break it for the `name` key of the output. Earlier, we used to put the text of the response directly in the `name` key but now since we want to support multimedia elements and multiple response templates, that definitely has to change.
But happy to hear how we can gracefully deprecate it maybe. @tmbo any thoughts?

erohmensing commented on changelog/6466.improvement.rst:
```diff
@@ -0,0 +1,6 @@
+[Response selector](retrieval-actions.mdx) is now trained on retrieval intent labels by default instead of actual response text.
```
This file should be `changelog/6466.improvement.md` and you no longer need the double backticking for code formatting!

erohmensing commented on changelog/6466.improvement.rst:
```diff
@@ -0,0 +1,6 @@
+If you want to use the actual response text for training, then use ``train_on_text=True`` in the configuration of ``ResponseSelector``.
+You can now also have multiple response templates for a single sub-intent of a retrieval intent. The first response template
+containing the text attribute is picked for training(if ``train_on_text=True``) and a random template is picked for bot's utterance just as how other ``utter_`` templates are picked.
+All response selector related evaluation artifacts - ``report.json, successes.json, errors.json, confusion_matrix.png`` now use the sub-intent of the retrieval intent instead of the actual response text.
```
```suggestion

All response selector related evaluation artifacts - ``report.json, successes.json, errors.json, confusion_matrix.png`` now use the sub-intent of the retrieval intent instead of the actual response text.
```

erohmensing commented on changelog/6466.improvement.rst:
```diff
@@ -0,0 +1,6 @@
+[Response selector](retrieval-actions.mdx) is now trained on retrieval intent labels by default instead of actual response text.
+
+If you want to use the actual response text for training, then use ``train_on_text=True`` in the configuration of ``ResponseSelector``.
+You can now also have multiple response templates for a single sub-intent of a retrieval intent. The first response template
```
```suggestion

You can now also have multiple response templates for a single sub-intent of a retrieval intent. The first response template
```

erohmensing commented on rasa/core/actions/action.py:
```diff
@@ -248,10 +249,17 @@ async def run(
-        message = selected[OPEN_UTTERANCE_PREDICTION_KEY]
-        message["template_name"] = selected[FULL_RETRIEVAL_INTENT]
+        possible_messages = selected[OPEN_UTTERANCE_PREDICTION_KEY]["name"]
 
-        return [create_bot_utterance(message)]
+        # Pick a random message from list of candidate messages
+        picked_message_idx = random.randint(0, len(possible_messages) - 1)
+        picked_message = copy.deepcopy(possible_messages[picked_message_idx])
```
I know before you were looking into using the NLG. I think here we don't support channel-specific utterances? That is something I think users will expect support for

dakshvar22 replied:
Hmm that's tricky. I would definitely like to make use of the NLG for it which means we need to have the response templates shared between the domain and NLU training data. I'll prefer to keep it out of scope of this PR till we have handled that. I'll mention it in the changelog though.

erohmensing replied:
Sounds good, creating an issue for it would be good too :)

tmbo commented on rasa/nlu/selectors/response_selector.py:
```diff
@@ -338,21 +349,31 @@ def _full_response(self, label: Dict[Text, Any]) -> Optional[Dict[Text, Any]]:
-                if hash(response.get(TEXT, "")) == label.get("id"):
-                    return response
-        return None
+            if self.train_on_text:
```
this function isn't only used training, this property should get a better name

dakshvar22 replied:
Changed it to `use_text_as_label`.

tmbo commented on rasa/nlu/selectors/response_selector.py:
```diff
@@ -338,21 +349,31 @@ def _full_response(self, label: Dict[Text, Any]) -> Optional[Dict[Text, Any]]:
+            if self.train_on_text:
+                for response in responses:
+                    if hash(response.get(TEXT, "")) == label.get("id"):
+                        return key, responses
```
why does this need to return a tuple, can't the caller also use `self.responses[key]`?

dakshvar22 replied:
Refactored the whole function to be cleaner.

tmbo commented on rasa/nlu/selectors/response_selector.py:
```diff
@@ -262,9 +267,9 @@ def _create_retrieval_intent_mapping(
-            retrieval_intent_mapping[
-                example.get(RESPONSE)
-            ] = f"{example.get(INTENT)}/{example.get(RESPONSE_KEY_ATTRIBUTE)}"
+            retrieval_intent_mapping[example.get(RESPONSE)] = example.get(
```
please add a function description of what the dict is that get's returned here

dakshvar22 replied:
Removed the usage of this function.

tmbo commented on rasa/nlu/selectors/response_selector.py:
```diff
@@ -262,9 +267,9 @@ def _create_retrieval_intent_mapping(
 
         retrieval_intent_mapping = {}
         for example in training_data.intent_examples:
```
should this rather loop over `training_data.response_examples`? otherwise there is always going to be 
`None: None` entry in this dictionary

tmbo commented on rasa/nlu/training_data/training_data.py:
```diff
@@ -243,15 +239,22 @@ def _fill_response_phrases(self) -> None:
+                    if assistant_utterance.get(TEXT) is not None:
+                        example.set(RESPONSE, assistant_utterance.get(TEXT))
+
+                # If none of the response templates have a text attribute, then use the key as the training label
```
this looks to long

tmbo commented on rasa/nlu/training_data/training_data.py:
```diff
@@ -243,15 +239,22 @@ def _fill_response_phrases(self) -> None:
+                # Use the first response text as training label if needed downstream
+                for assistant_utterance in assistant_utterances:
+                    if assistant_utterance.get(TEXT) is not None:
+                        example.set(RESPONSE, assistant_utterance.get(TEXT))
```
```suggestion
                        example.set(RESPONSE, assistant_utterance[TEXT])
```

tmbo commented on rasa/nlu/training_data/training_data.py:
```diff
@@ -243,15 +239,22 @@ def _fill_response_phrases(self) -> None:
+
+                # Use the first response text as training label if needed downstream
+                for assistant_utterance in assistant_utterances:
+                    if assistant_utterance.get(TEXT) is not None:
```
```suggestion
                    if assistant_utterance.get(TEXT):
```
I don't think we should train on `""`

tmbo commented:
I am currently reviewing this as well

tmbo commented on rasa/nlu/selectors/response_selector.py:
```diff
@@ -338,21 +349,31 @@ def _full_response(self, label: Dict[Text, Any]) -> Optional[Dict[Text, Any]]:
         for ranking in label_ranking:
-            ranking["full_retrieval_intent"] = self.retrieval_intent_mapping.get(
-                ranking.get("name")
+            ranking["full_retrieval_intent"] = (
```
is a `full_retrieval_intent ` the same as `example.get_combined_intent_response_key`? if so please unify the naming

tmbo commented on rasa/nlu/selectors/response_selector.py:
```diff
@@ -338,21 +349,31 @@ def _full_response(self, label: Dict[Text, Any]) -> Optional[Dict[Text, Any]]:
-            ranking["full_retrieval_intent"] = self.retrieval_intent_mapping.get(
-                ranking.get("name")
+            ranking["full_retrieval_intent"] = (
+                self.retrieval_intent_mapping.get(ranking.get("name"))
```
why do we need to make this differentiation?

tmbo commented on rasa/nlu/selectors/response_selector.py:
```diff
@@ -366,9 +387,13 @@ def process(self, message: Message, **kwargs: Any) -> None:
-            "response": self._full_response(label) or {TEXT: label.get("name")},
+            "response": {
+                "id": label["id"],
+                "name": label_responses,
```
why should this be called "name" ?

dakshvar22 replied:
Changed it to `response_templates`

tmbo commented on tests/nlu/selectors/test_selectors.py:
```diff
@@ -66,3 +69,60 @@ def test_train_selector(pipeline, component_builder, tmpdir):
         assert rank.get("full_retrieval_intent") is not None
+
+
+def test_training_label():
```
ideally, this name should tell the reader what this is testing

tmbo commented on tests/nlu/selectors/test_selectors.py:
```diff
@@ -66,3 +69,60 @@ def test_train_selector(pipeline, component_builder, tmpdir):
+        list(td.responses.keys())
+    )
+
+    rs = ResponseSelector(component_config={TRAIN_ON_TEXT: True})
```
this looks very much like a separate test

tmbo commented on tests/nlu/selectors/test_selectors.py:
```diff
@@ -66,3 +69,60 @@ def test_train_selector(pipeline, component_builder, tmpdir):
+def test_training_label():
+
+    # use data that include some responses
+    td = load_data("data/examples/rasa/demo-rasa.md")
```
please use more readable names (`td`, `rs`, `r`). and please don't reassign them

tmbo commented on tests/nlu/selectors/test_selectors.py:
```diff
@@ -66,3 +69,60 @@ def test_train_selector(pipeline, component_builder, tmpdir):
+    rs.preprocess_train_data(td)
+
+    assert rs.responses == td.responses
+    assert sorted(list(rs.index_label_id_mapping.values())) == sorted(
```
i don't know what this tests

tmbo commented on tests/nlu/selectors/test_selectors.py:
```diff
@@ -66,3 +69,60 @@ def test_train_selector(pipeline, component_builder, tmpdir):
+
+
+@pytest.mark.parametrize(
+    "pred_label, train_on_text, full_response_label, full_response_text",
```
what is a `full_response_label` was that not previously a `full_response_key` or `intent_response_key` or `combined_intent_response_key`. 

the naming is quite confusing:
- what is the name of `faq/ask_name`?
- what is the name of `faq`?
- what is the name of `ask_name`?

dakshvar22 replied:
So, `faq` is the `retrieval intent` or just `intent`(depending on the context), `ask_name` is the `response_key` and `faq/ask_name` is the `intent_response_key`.
I'll change the names in the parameters here to better reflect this naming.

tmbo commented on tests/nlu/selectors/test_selectors.py:
```diff
@@ -66,3 +69,60 @@ def test_train_selector(pipeline, component_builder, tmpdir):
+
+
+@pytest.mark.parametrize(
+    "pred_label, train_on_text, full_response_label, full_response_text",
```
`faq/ask_name` also doesn't sound like a `full_response_text`

tmbo reviewed (changes requested):
thanks a lot for fixing the bug in the response selector and for implementing multi template support for responses. 

I think the code would benefit from a better naming scheme: even after working on the responses, I find the naming very confusing in the code base. e.g. it is not clear to me what the difference between `INTENT_RESPONSE_KEY` and `FULL_RETRIEVAL_INTENT` is.

dakshvar22 commented on rasa/nlu/selectors/response_selector.py:
```diff
@@ -249,6 +253,7 @@ def model_class() -> Type[RasaModel]:
 
     def _load_selector_params(self, config: Dict[Text, Any]) -> None:
         self.retrieval_intent = config[RETRIEVAL_INTENT]
+        self.train_on_text = config[TRAIN_ON_TEXT]
```
The call to `_load_selector_params()` is in `_check_config_parameters` which is called from the constructor of the parent class. So the property is defined inside init. Would you recommend changing it still?

tmbo replied:
a property should always be initialised in `__init__` (a lot of tools depend on this).

it is fine to set the value here, but `def __init__` should have a `self.train_on_text = None` in that case (or some other default value like `False`)

dakshvar22 commented on changelog/6466.improvement.rst:
```diff
@@ -0,0 +1,6 @@
+[Response selector](retrieval-actions.mdx) is now trained on retrieval intent labels by default instead of actual response text.
+
+If you want to use the actual response text for training, then use ``train_on_text=True`` in the configuration of ``ResponseSelector``.
```
Already in the migration guide - https://github.com/RasaHQ/rasa/pull/6466/files#diff-a6ba427077dfb18d7f8aea2d50f40a34

dakshvar22 commented:
@tmbo I agree that the naming has got very confusing. Some of them I couldn't change so that some of the existing keys that codes outside of Rasa(example - Rasa X, Sara) could rely on doesn't change. Does it make sense to make the naming more consistent once and for all since this goes into a major release? Related to things like the confusing keys - "name", "full_retrieval_intent", INTENT_RESPONSE_KEY, etc.

tmbo commented:
yes, the naming should be more consistent, and as far as I can see this is also not related to the API, since you changed that to a dict anyways and moved the properties. I'd rather add that deprecated information in a separate call using the old keys just in there than have confusing variable and function names all over the place. 

I don't want this to sound overly picky, but working on this and trying to implement the multimedia responses just showed how hard it was for me to understand the response selector code.

dakshvar22 commented:
@tmbo Thanks for the feedback. I am working on cleaning/simplifying(including naming conventions) as much code as I can for the response selector.

dakshvar22 commented:
@tmbo The PR is ready for another review. I have made the following changes since the last review - 
1. I have tried to stick to the following naming scheme - If the intent name appears as `faq/ask_name` in the training data, then the message object should have the following attributes - 
`intent` = `faq`
`response_key` = `ask_name`
`intent_response_key` = `faq/ask_name`
`response` = All the response templates under this `faq/ask_name`
1. Deprecate `full_retrieval_intent` completely from the output of the `ResponseSelector` and instead use `intent_response_key` to store the full intent name as it appears in the training data. This is more consistent with the naming scheme used in the code.
2. Deprecate `name` as the key to store the predicted response text and instead use `response_templates` to output all available response templates for the predicted `intent_response_key`. The key `name` makes more sense for the output of intent classifiers and I had just used that earlier. My bad.
3. Change the schema of `ResponseSelectionEvaluationResult` to only have [`intent_response_key_target`, `intent_response_key_prediction`, `message`, `confidence`] as its attributes. Since we are now dealing with multiple response texts for each `intent_response_key`, it did not make sense to have the `response_target`(which used to hold the text of the response) in there as the match is only based on keys now and not on the actual text.
4. Deprecate `get_combined_intent_response_key` method from `Message` class since `intent_response_key` essentially holds that already. There is a new method `get_full_intent` which returns either `intent_response_key` attribute(if it exists) or `intent` attribute.
5. Change the new parameter name in `ResponseSelector` from `train_on_text` to `use_text_as_label`.
6. A couple of model level changes to simplify which exact implementation of DIET is used in different cases depending on `use_text_as_label=True/False`.


A couple of questions that I still have which could simplify the code further maybe - 
1. For 1&2 I am not sure how to implement a graceful deprecation for them because any client code could still be querying these keys. What's the best way to do that?
2. [This part of the code](https://github.com/RasaHQ/rasa/pull/6466/files#diff-75686dc39a842033808ce0c6137c102bR363-R372) seems overly complex because of the `or` construct that's needed as `_resolve_intent_response_key` can theoretically return `None`. Practically, if that happens there is already a problem in the training data which the user is warned about [here](https://github.com/RasaHQ/rasa/pull/6466/files#diff-0e775499654bc8adf9290d502dac3c0dR427-R433). Do you recommend getting around this in some other way?

erohmensing commented on docs/docs/migration-guide.mdx:
```diff
@@ -341,6 +341,12 @@ Please see [Forms](./forms.mdx) if you have further customizations in your
 
 <a aria-hidden="true" tabIndex="-1" className="anchor enhancedAnchor" id="migration-to-rasa-1-8"></a>
 
+### Response Selectors
+
+Response Selectors are now trained on retrieval intent labels by default instead of the actual response text. For most models, this should improve training time and accuracy of the `ResponseSelector`.
+
+If you want to revert to the pre-2.0 default behavior, add the `use_text_as_label=true` parameter to your `ResponseSelector` component.
+
```
The output schema/api change should be in here too!

dakshvar22 replied:
I just added it :)

dakshvar22 replied:
Ohh, you meant in the migration guide. Adding it now.

dakshvar22 commented on rasa/nlu/selectors/response_selector.py:
```diff
@@ -244,30 +256,20 @@ def label_key(self) -> Text:
     @staticmethod
-    def model_class() -> Type[RasaModel]:
-        return DIET2DIET
+    def model_class(use_text_as_label: bool) -> Type[RasaModel]:
+        if use_text_as_label:
+            return DIET2DIET
+        else:
+            return DIET2BOW
```
This and the almost the whole of `DIET2BOW` class was needed because metrics are tightly coupled with the model architecture right now. See https://github.com/RasaHQ/research/issues/112

dakshvar22 commented on rasa/nlu/selectors/response_selector.py:
```diff
@@ -204,6 +210,9 @@ def required_components(cls) -> List[Type[Component]]:
         RETRIEVAL_INTENT: None,
+        # Boolean flag to check if actual text of the response
+        # should be used as ground truth label for training the model.
+        USE_TEXT_AS_LABEL: True,
```
Reminder for myself: Change this to False before merging. This is `True` for model running model regression tests once.

tmbo commented on docs/docs/migration-guide.mdx:
````diff
@@ -341,6 +341,75 @@ Please see [Forms](./forms.mdx) if you have further customizations in your
+ the target and predicted labels instead of the actual response text.
+This means that the output in these files should look like -
+
+`response_selection_successes.json`
+```json
````
```suggestion
```json title="response_selection_successes.json"
```

tmbo commented on docs/docs/migration-guide.mdx:
````diff
@@ -341,6 +341,75 @@ Please see [Forms](./forms.mdx) if you have further customizations in your
+}
+```
+
+`response_selection_errors.json`
+```json
````
```suggestion
```json title="response_selection_errors.json"
```

tmbo commented on examples/responseselectorbot/credentials.yml:
```diff
@@ -0,0 +1,32 @@
+twilio:
```
I don't think these should be in here

dakshvar22 replied:
@tmbo Do you mean the actual credentials shouldn't be in here and we should put some dummy credentials?
I actually assumed these are dummy credentials that don't work because I took them from the moodbot example. We should change it there too if these are actually working credentials.

dakshvar22 replied:
I see what you mean. We are not using channels in the example, so the file `credentials.yml` isn't needed even.

tmbo commented on examples/responseselectorbot/data/responses.yml:
```diff
@@ -0,0 +1,12 @@
+  chitchat/ask_weather:
+  - text: Oh, it does look sunny right now in Berlin.
+    image: "https://i.imgur.com/vwv7aHN.png"
+  - text: I am not sure of the whole week but I can see the sun is out today.
```
```suggestion
  - text: I am not sure of the whole week but I can see the sun is out today.

```

tmbo commented on tests/core/test_actions.py:
```diff
@@ -363,10 +370,12 @@ async def test_action_utter_retrieved_response(
         "Who are you?",
         parse_data={
-            "response_selector": {
+            RESPONSE_SELECTOR_PROPERTY_NAME: {
```
optional: I am not sure we should extensively use constants in the tests:
- it makes them less readable
- it tightly couples the test to the implementation: the test simulates an API user, if we use the same constants here as we do in the code, it is less obvious that changing that constant breaks that API (as the test will still pass)

dakshvar22 replied:
Makes sense! I'll change them back to the actual strings.
